browser/brave_shell_integration_win.cc

```diff
@@ -0,0 +1,170 @@
+    ModifyFunc* Modify;    // 16
+  };
+
+  struct ILFreeDeleter {
+    void operator()(LPITEMIDLIST ptr) {
+      if (ptr) {
+        ILFree(ptr);
+      }
+    }
+  };
+
+  IPinnedList3* pinned_list = nullptr;
+  HRESULT hr = CoCreateInstance(CLSID_TaskbandPin, nullptr,
+                                CLSCTX_INPROC_SERVER, IID_IPinnedList3,
+                                reinterpret_cast<void**>(&pinned_list));
+  if (FAILED(hr) || !pinned_list) {
+    LOG(ERROR) << __func__ << " CoCreateInstance failed.";
+    return;
+  }
+
+  std::unique_ptr<__unaligned ITEMIDLIST, ILFreeDeleter> path(
+      ILCreateFromPathW(shortcut_path.value().c_str()));
+  hr =
+      pinned_list->vtbl->Modify(pinned_list, nullptr, path.get(), PLMC_INT_MAX);
+  pinned_list->vtbl->Release(pinned_list);
+
+  if (FAILED(hr))
+    LOG(ERROR) << __func__ << " Failed to pin to taskbar.";
+}
+
+void PinToTaskbarImpl() {
+  base::FilePath chrome_exe;
+  if (!base::PathService::Get(base::FILE_EXE, &chrome_exe))
+    return;
+
+  ShellUtil::ShortcutProperties properties(ShellUtil::CURRENT_USER);
+  ShellUtil::AddDefaultShortcutProperties(chrome_exe, &properties);
+  // Generate the shortcut path and ask to pin it.
+  base::FilePath shortcut_path;
+  ShellUtil::GetShortcutPath(ShellUtil::SHORTCUT_LOCATION_DESKTOP,
+                             ShellUtil::CURRENT_USER, &shortcut_path);
+  std::wstring shortcut_name(ExtractShortcutNameFromProperties(properties));
+  shortcut_path = shortcut_path.Append(shortcut_name);
+
+  bool create_shortcut = true;
+  if (!base::PathExists(shortcut_path))
+    create_shortcut = false;
+
+  if (create_shortcut && !ShellUtil::CreateOrUpdateShortcut(
+                             ShellUtil::SHORTCUT_LOCATION_DESKTOP, properties,
+                             ShellUtil::SHELL_SHORTCUT_CREATE_ALWAYS)) {
+    LOG(ERROR) << __func__ << " Failed to create shortcut.";
+    return;
+  }
+
+  PinToTaskbarWin10(shortcut_path);
+}
+
+void OnShellHandlerConnectionError() {
+  // Do nothing when checking failed.
+}
+
+void OnIsPinnedToTaskbarResult(bool succeeded,
+                               bool is_pinned_to_taskbar,
+                               bool is_pinned_to_taskbar_verb_check) {
+  if (!succeeded)
+    return;
+
+  if (is_pinned_to_taskbar || is_pinned_to_taskbar_verb_check)
+    return;
+
+  base::ThreadPool::CreateCOMSTATaskRunner({base::MayBlock()})
+      ->PostTask(FROM_HERE, base::BindOnce(&PinToTaskbarImpl));
+}
+
+}  // namespace
+
+namespace shell_integration::win {
+
+// TODO(simonhong): Support profile specific shortcut pinning.
+// For now, only default profile's shortcut is created and pinned.
+void PinToTaskbar() {
+  // According to the FF, it's only supported from specific win10 version.
+  // See PinCurrentAppToTaskbarAsyncImpl() at
+  // https://github.com/mozilla/gecko-dev/blob/master/browser/components/shell/nsWindowsShellService.cpp
+  if (base::win::GetVersion() < base::win::Version::WIN10_RS5)
+    return;
```

Comment:
Then, can we try something like this?

```suggestion
  if (base::win::CanPinShortcut()) {
       ...// call ShellExecute version base::win::PinShortcutToTaskbar() in callback
  }
     return;
```

Reply:
Test with upstream methods (IsPinned/Pin) works nicely. Copied them used.
We can delete copied one when it's available.

Reply:
`base::win::CanPinShortcutToTaskbar()` uses different condition with this one.
`base::win::GetVersion() < base::win::Version::WIN10` vs.
`base::win::GetVersion() >= base::win::Version::WIN10_RS5`

Reply:
Yeah, so what i mean is

```
if (base::win::GetVersion() >= WIN10_RS5)
  PinTaskBar with IPinnedList3 - ShellUtill's implemntation
else if (base::win::GetVersion() < base::win::Version::WIN10)
  PinTaskBar with ShellExecute() - Shorcut's implementation.
```

so that we can cover more versions?

Reply:
Ok, I see what you mean. How about handling it as a next step?
As this PR is for pinning from first run dialog, we don't need to care about win7/8.
The reason is windows installer tries to pin to taskbar and it works on win7/8. So, don't need to try to pin again on Win7/8. I planned to handle it as a next task for pinning when user set Brave as a default from settings menu.

Reply:
Oh, sure! Fair enough

Reply:
@sangwoo108 Thanks for review! 👍🏼